### PR TITLE
Bind in CellParams for RNN

### PR DESCRIPTION
**Pull request by jamesr66a**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #34140 [quantization] Use torchbind for Linear and RNN PackedParams
* **#35787 Bind in CellParams for RNN**

Differential Revision: [D20784118](https://our.internmc.facebook.com/intern/diff/D20784118)

### CODE REVIEW

File: aten/src/ATen/native/RNN.cpp [hunk 1]
```diff
@@ -1182,20 +1670,37 @@ Tensor rnn_relu_cell(
...
+  params.reserve(_params_.size());
+  for (const auto& param : _params_) {
+    params.emplace_back(static_cast<c10::intrusive_ptr<CellParamsBase>>(param));
+  }
   TORCH_CHECK(hx.size() == 2, "lstm expects two hidden states");
   if (at::cudnn_is_acceptable(_input)) {
     Tensor output, hy, cy;
-    lstm_cudnn_stub(_input.device().type(), output, hy, cy, _input, hx, _params, has_biases,
-                    num_layers, dropout_p, train, bidirectional, batch_first);
+    // TODOJAMES
```
Comment (dzhulgakov) [inline aten/src/ATen/native/RNN.cpp:1694]: seems like this is not done yet?

**Review by dzhulgakov — COMMENTED** [review]
RNNs are magical

File: aten/src/ATen/native/RNN.cpp (hunk 1 above)
Comment (jamesr66a) [reply aten/src/ATen/native/RNN.cpp:1694]: I think this whole part should be deleted, afaik CUDNN never supported quantized LSTMs

File: aten/src/ATen/core/jit_type.h [hunk 2]
```diff
@@ -1386,6 +1386,12 @@ struct getTypePtr_<int64_t> final {
   }
 };
 template <>
+struct getTypePtr_<c10::ScalarType> final {
+  static TypePtr call() {
+    return IntType::get();
```
Comment (dzhulgakov) [inline aten/src/ATen/core/jit_type.h:1391]: wait, what is it? isn't scalar like a union of int and float?

File: aten/src/ATen/native/RNN.cpp (hunk 1 above)
Comment (dzhulgakov) [reply aten/src/ATen/native/RNN.cpp:1694]: then delete it and add TORCH_INTERNAL_ACCEPT(false) or just kill the entire conditional

File: aten/src/ATen/native/RNN.cpp [hunk 3]
```diff
@@ -1182,20 +1670,37 @@ Tensor rnn_relu_cell(
...
+    // lstm_cudnn_stub(_input.device().type(), output, hy, cy, _input, hx,
+    // _params, has_biases,
+    //                 num_layers, dropout_p, train, bidirectional,
+    //                 batch_first);
     return std::make_tuple(std::move(output), std::move(hy), std::move(cy));
   }
   auto result_dtype = dtype.has_value() ? dtype.value() : at::kChar;
-  check_device(_input, _params, hx);
+  // TODOJAMES
+  // check_device(_input, _params, hx);
```
Comment (dzhulgakov) [inline aten/src/ATen/native/RNN.cpp:1703]: even if it's always cpu - either just comment out or delete

**Review by dzhulgakov — APPROVED** [review]
Ok, I didn't read each and every line but overall flow seems reasonable and I'll trust you on BC concerns.

File: aten/src/ATen/core/jit_type.h (hunk 2 above)
Comment (jamesr66a) [reply aten/src/ATen/core/jit_type.h:1391]: Yeah you're right, I think this should be `ScalarTypeType::get()`

Comment (jamesr66a) [reply aten/src/ATen/core/jit_type.h:1391]: Looks like I have to fix up other existing erroneous sites also, e.g. https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/frontend/schema_type_parser.cpp#L39

Comment (eellison) [reply aten/src/ATen/core/jit_type.h:1391]: You're thinking of `Scalar` which is `NumberType::get()`

Comment (jamesr66a) [reply aten/src/ATen/core/jit_type.h:1391]: also, this is `ScalarType`, not `Scalar`, lol this is confusing

**Pull request closed without merging**